### PR TITLE
mds: add debug logs during setxattr ceph.dir.subvolume

### PR DESCRIPTION
**Pull request by mchangir**
* dump the subvolume flag for inodes
* add debug logs when doing setxattr for ceph.dir.subvolume

Fixes: https://tracker.ceph.com/issues/61958
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/CInode.cc [hunk 1]
```diff
@@ -5165,6 +5165,11 @@ void CInode::dump(Formatter *f, int flags) const
     }
     f->close_section();
   }
+
+  SnapRealm *realm = find_snaprealm();
```
Comment (vshankar) [inline src/mds/CInode.cc:5169]: minor nit: use `auto` (you can choose to ignore this comment)

File: src/mds/Server.cc [hunk 2]
```diff
@@ -6192,6 +6192,10 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
       inodeno_t subvol_ino = realm->get_subvolume_ino();
       // can't create subvolume inside another subvolume
       if (subvol_ino && subvol_ino != cur->ino()) {
+	dout(20) << "subvol ino changed between rdlock release and xlock "
+		 << "policylock; subvol_ino:" << subvol_ino << " "
+		 << "cur->ino:" << cur->ino()
```
Comment (dparmar18) [inline src/mds/Server.cc:6197]: need space after colons

File: src/mds/Server.cc [hunk 3]
```diff
@@ -6206,10 +6210,13 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
     auto pi = cur->project_inode(mdr, false, true);
     if (!srnode)
       pi.snapnode->created = pi.snapnode->seq = realm->get_newest_seq();
-    if (val)
+    if (val) {
+      dout(20) << "marking subvolume for ino:" << cur->ino() << dendl;
       pi.snapnode->mark_subvolume();
-    else
+    } else {
+      dout(20) << "clearing subvolume for ino:" << cur->ino() << dendl;
```
Comment (dparmar18) [inline src/mds/Server.cc:6217]: here too

File: src/mds/CInode.cc [hunk 4]
```diff
@@ -5165,6 +5165,11 @@ void CInode::dump(Formatter *f, int flags) const
     }
     f->close_section();
   }
+
+  auto realm = find_snaprealm();
+  inodeno_t subvol_ino = realm->get_subvolume_ino();
+  bool is_subvol = (subvol_ino && subvol_ino == ino());
```
Comment (dparmar18) [inline src/mds/CInode.cc:5171]: `subvol_ino && (subvol_ino == ino())` be better right

File: src/mds/Server.cc (hunk 2 above)
Comment (mchangir) [reply src/mds/Server.cc]: added space

File: src/mds/Server.cc [hunk 5]
```diff
@@ -6206,10 +6210,13 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
     auto pi = cur->project_inode(mdr, false, true);
     if (!srnode)
       pi.snapnode->created = pi.snapnode->seq = realm->get_newest_seq();
-    if (val)
+    if (val) {
+      dout(20) << "marking subvolume for ino:" << cur->ino() << dendl;
```
[the remark this answers is not in the document]
Comment (mchangir) [reply src/mds/Server.cc]: added space

File: src/mds/Server.cc (hunk 3 above)
Comment (mchangir) [reply src/mds/Server.cc]: added space

**Comment by mchangir** [comment]
jenkins test make check arm64

File: src/mds/CInode.cc (hunk 4 above)
Comment (mchangir) [reply src/mds/CInode.cc:5171]: The equality operator has higher precedence than the logical AND operator as per [C++ Operator Precedence Rules](https://en.cppreference.com/w/cpp/language/operator_precedence).

Comment (dparmar18) [reply src/mds/CInode.cc:5171]: in that case do we really need the outer parenthesis? `bool is_subvol = subvol_ino && subvol_ino == ino();` would suffice

**Comment by mchangir** [comment]
@vshankar https://tracker.ceph.com/projects/cephfs/wiki/Main

**Review by vshankar — APPROVED** [review]

**Comment by mchangir** [comment]
http://pulpito.front.sepia.ceph.com/mchangir-2023-09-12_05:40:22-fs-wip-mchangir-testing-20230908.140927-testing-default-smithi/

**Pull request merged**